Asap Client Authentication Extension.

jamesmoessis opened the pull request:
**Description:**
Adding an extension which allows the collector to authenticate outgoing requests with ASAP ([Atlassian Service Authentication Protocol](https://s2sauth.bitbucket.io/)). 

It works by implementing `ClientAuthenticator` and setting the authorization header as the encoded JWT. It uses the [go-asap](https://bitbucket.org/atlassian/go-asap/src) library to achieve much of its functionality.

**Link to tracking Issue:**
~~N/A~~  https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5930

**Testing:** 
We mock key server and a base RoundTripper to achieve validation on signature and fields on the JWT. Configuration and factory are also tested. There are yaml files for good and bad config (one good, multiple bad configurations). 

Additionally, at Atlassian we have been using ASAP in the collector for a long time already.

**Documentation:** 

README.md gives a sample configuration and brief description. Those familiar with ASAP should have little trouble setting it up 

Additionally, @MovieStoreGuy and myself are happy to be the codeowners for this extension.

jamesmoessis committed:
initial commit for asap client auth

jamesmoessis committed:
format

jamesmoessis committed:
test for sstandard asap client config

jamesmoessis committed:
tests for factory god + bad config

jamesmoessis committed:
implement PerRPCCredentials()

jamesmoessis committed:
tests for main functionality among other small changes

jamesmoessis committed:
update readme

jamesmoessis committed:
appease linter

jamesmoessis committed:
Update extension/asapclientauthextension/config.go

Co-authored-by: Sean Marciniak <30928402+MovieStoreGuy@users.noreply.github.com>

jamesmoessis committed:
implement some of sean's review

jamesmoessis committed:
Update extension/asapclientauthextension/config_test.go

Co-authored-by: Sean Marciniak <30928402+MovieStoreGuy@users.noreply.github.com>

jamesmoessis committed:
implement more review

jamesmoessis committed:
implement minor review

jamesmoessis committed:
appease linter

MovieStoreGuy reviewed (approved):
I had approved this on our fork of the collector.

MovieStoreGuy commented:
The related issue is https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5930

jamesmoessis committed:
add  CODEOWNERS for asapauth

Commit:
Updating the ASAP extention

There was a few parts missing from the extention. Updated the import to
match directory name, added it to the mod file and version, then to the
exported extensions.

Fixing test case

MovieStoreGuy commented:
Going to collapse my two commits into one in order to rerun the failing tests.

mx-psi commented on go.mod:
```diff
@@ -526,6 +532,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elast
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/zipkinexporter => ./exporter/zipkinexporter
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth => ./extension/asapauth
```
This should be done on a separate PR; read the [contributing doc for new components](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component)

jamesmoessis replied:
Thanks, I've removed all the changes to the root `go.mod`. However, the build fails saying "asapauth/go.mod is not listed in any module set". Since I can't list it in the root go.mod, is there something I can do to make this green? Or is it fine to leave it like this.

mx-psi replied:
You need to add it on `versions.yaml` (we should add this step to the contributing guidelines, sorry, it's confusing)

mx-psi commented on internal/components/components.go:
```diff
@@ -130,6 +131,7 @@ func Components() (component.Factories, error) {
 	delete(factories.Extensions, "pprof")
 
 	extensions := []component.ExtensionFactory{
+		asapauth.NewFactory(),
```
ditto

jamesmoessis replied:
Removed

mx-psi commented on extension/asapauth/go.mod:
```diff
@@ -0,0 +1,47 @@
+	bitbucket.org/atlassian/go-asap v0.0.0-20210823164515-3d5971c40c4c
+	github.com/SermoDigital/jose v0.9.2-0.20161205224733-f6df55f235c2
+	github.com/stretchr/testify v1.7.0
+	go.opentelemetry.io/collector v0.38.0
```
This should match the version on the `main` branch

jamesmoessis replied:
Fixed!

mx-psi commented on extension/asapauth/factory.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+func createExtension(_ context.Context, _ component.ExtensionCreateSettings, cfg config.Extension) (component.Extension, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
```
Validation [will be done by the service pipeline during initialization](https://github.com/open-telemetry/opentelemetry-collector/blob/bdcb989da369fd2b06be71966dd6f3259aca7fdc/service/service.go#L46-L48), so there is no need to do this again here.

```suggestion
```

mx-psi commented on extension/asapauth/config.go:
```diff
@@ -0,0 +1,62 @@
+	PrivateKey string `mapstructure:"private_key"`
+}
+
+func (c *Config) Validate() error {
```
Maybe you want to use `multierr` to return all the errors here instead of the first one?

jamesmoessis replied:
Done!

jamesmoessis committed:
Update extension/asapauth/factory.go

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

jamesmoessis committed:
use multierr for config validation

jamesmoessis committed:
Revert "Updating the ASAP extention"

This reverts commit 7a8cfd513dce5cff20e844fd3b70096c63e33cba.

jamesmoessis committed:
correct module name

jamesmoessis committed:
fix test

jamesmoessis committed:
go mod tidy

mx-psi commented:
Looks good once added to the `versions.yaml` file!

jamesmoessis committed:
Merge branch 'main' into asap-auth-extension

jamesmoessis committed:
add asapauth to versions.yaml

jamesmoessis committed:
empty commit to trigger build

jamesmoessis committed:
empty commit to trigger build

jamesmoessis commented:
Thanks @mx-psi, I've updated the `versions.yaml`. I also raised a PR to add it to the contributing guide: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6411

The tests seem pretty flakey right now. I can't seem to get it to go green, and the failures are unrelated to this PR.

jamesmoessis commented:
Looks like the build passed on my 2nd empty commit 😀

jamesmoessis committed:
Merge branch 'main' into asap-auth-extension

jamesmoessis committed:
bring up to date with collector 0.40.0

jamesmoessis commented:
Hey @mx-psi, any idea when this will be merged?

jamesmoessis committed:
attempt to appease lint

mx-psi commented:
@jamesmoessis I was waiting on @tigrannajaryan to review to mark as `ready-to-merge`, let's give them a couple days since there were some holiday days on the US and after that I will mark this since I see no blockers

jamesmoessis commented:
Awesome thankyou 😄

tigrannajaryan commented:
@jpkrohling you are the auth expert, can you please review?

zenmoto committed:
move MongoDB Atlas receiver into components (#6367)

* move MongoDB Atlas receiver into components

* added receiver to receivers_test

* add changelog entry

bertysentry committed:
issue #6142 Fixed prometheusexecreceiver command line processing on Windows (#6145)

Fixed execution of commands in `prometheusexecreceiver` on Windows

**Link to tracking Issue:** Issue #6142

Commit:
`dependabot`: updates as per dependabot (#6458)

Bump github.com/DataDog/agent-payload/v5 from 5.0.2 to 5.0.3 in /exporter/datadogexporter
Bump github.com/DataDog/datadog-agent/pkg/quantile from 0.32.0 to 0.32.1 in /exporter/datadogexporter
Bump github.com/aliyun/aliyun-log-go-sdk from 0.1.24 to 0.1.27 in /exporter/alibabacloudlogserviceexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /exporter/awsemfexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /exporter/awskinesisexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /exporter/awsprometheusremotewriteexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /exporter/awsxrayexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /exporter/datadogexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /exporter/kafkaexporter
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /extension/observer/ecsobserver
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /internal/aws/awsutil
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /internal/aws/xray
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /internal/aws/xray/testdata/sampleapp
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /processor/resourcedetectionprocessor
Bump github.com/aws/aws-sdk-go from 1.42.9 to 1.42.14 in /receiver/awsxrayreceiver
Bump github.com/hashicorp/go-hclog from 0.16.2 to 1.0.0 in /exporter/logzioexporter
Bump github.com/influxdata/line-protocol/v2 from 2.2.0 to 2.2.1 in /receiver/influxdbreceiver
Bump github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common from 1.0.296 to 1.0.301 in /exporter/tencentcloudlogserviceexporter
Bump go.mongodb.org/atlas from 0.13.0 to 0.14.0 in /receiver/mongodbatlasreceiver

mustafain117 committed:
Add Prometheus Receiver tests for honor_labels configuration (#6369)

* refactor existing external labels test

* Add test for label_limit configuration

* shutdown mock prometheus http test server within testHelper

* fix spelling

* fix lint error

* rename testHelper to testMetricsReceiver

* fix: lint issue in metrics_receiver_helper_test.go

* Add Prometheus receiver test for honor_labels configuration

* update TestHonorLabelsFalseConfig to use testComponent

mustafain117 committed:
Add Prometheus Receiver test for label_value_length_limit configuration (#6430)

* Add Prometheus Receiver test for label_value_length_limit config

* rename target1 to targetLabelLimits1

mustafain117 committed:
Add Prometheus Receiver tests to run OpenMetrics test cases (#6409)

* Add Prometheus receiver test to run OpenMetrics test cases

* fix: resolved gosec lint issue

* fix error handling in getTestCase

* remove http requests in tests by copying openmetrics data into testdata

* only skip failing openmetrics tests using a deny-list

TomRoSystems committed:
Add setStatus operation for spanprocessor (#5886)

* Add setStatus operation for spanprocessor

* Change code to have text format instead of numeric and add another example

* Add "Unset" value and tests

* Update README for Unset value

* Apply suggestions from code review

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

* From review: Move statuses names to constants and fix "Err" => "Error"

* Code review: fixing style

* One more "Err" and fixing comment(s) as suggested from review

* Clear status description for cases different than error

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

dmitryax committed:
[receivers/hostmetrics] Fix set of attributes for system.cpu.time metric (#6422)

gbbr committed:
use otel.trace_id instead of otlp.trace_id (#6432)

dmitryax committed:
[receiver/mongoatlas] Fix attributes fields in metadata.yaml (#6440)

jpkrohling commented on extension/asapauth/README.md:
```diff
@@ -0,0 +1,30 @@
+# ASAP Client Authentication Extension
+
+This extension provides [Atlassian Service Authentication Protocol](https://s2sauth.bitbucket.io/) (ASAP) client 
+credentials for HTTP or RPC based exporters. 
```
```suggestion
credentials for HTTP or gRPC based exporters. 
```

jpkrohling commented on .github/CODEOWNERS:
```diff
@@ -53,6 +53,7 @@ extension/httpforwarder/                             @open-telemetry/collector-c
 extension/observer/                                  @open-telemetry/collector-contrib-approvers @asuresh4 @jrcamp
 extension/bearertokenauthextension/                  @open-telemetry/collector-contrib-approvers @jpkrohling @pavankrish123
 extension/oauth2clientauthextension/                 @open-telemetry/collector-contrib-approvers @pavankrish123
+extension/asapauth                                   @open-telemetry/collector-contrib-approvers @jamesmoessis @MovieStoreGuy
```
Other auth extensions use the suffix `authextension`. I would prefer to have this one follow the same pattern.

jamesmoessis replied:
I've changed the name to `asapauthextension`

jpkrohling commented on extension/asapauth/extension.go:
```diff
@@ -0,0 +1,98 @@
+	return nil
+}
+
+var _ configauth.ClientAuthenticator = (*AsapClientAuthenticator)(nil)
```
Perhaps move this closer to the type definition?

jamesmoessis replied:
Done 👍

jpkrohling commented on extension/asapauth/extension.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+func (a AsapClientAuthenticator) PerRPCCredentials() (credentials.PerRPCCredentials, error) {
+	token, err := a.provisioner.Provision()
```
IIRC, this is called only once, during the setup of the extension. Your `PerRPCAuth` type would need to handle the caching there, not here.

jamesmoessis replied:
Thanks for pointing this out, you are absolutely right. I have moved the token provisioning logic into `GetRequestMetadata()` in `PerRPCAuth`.

jpkrohling commented on extension/asapauth/README.md:
```diff
@@ -0,0 +1,30 @@
+      - someservice
+      - someotherservice
+    private_key: ${ASAP_PRIVATE_KEY}
+    ttl_seconds: 120 # Default: 60
```
What does this property do? Will a new token be obtained after the TTL? Would an existing token just be validated after the TTL? If this is explained in another authoritative document, like Asap's documentation, link there.

jamesmoessis replied:
The `exp` (expiration) claim on the JWT is set as `time.Now().Add(TTL)`. So, the token will be cached until it's expired, in which case another token will be provisioned. 

The client caching itself is implementation specific, but is related to what the spec says about token expiry, so I have linked that information here. Additionally, I have added comments to all of these yaml fields with a brief on what they are.

Aneurysm9 committed:
Fix/broken main (#6461)

* processor/spanprocessor: fix attribute assignment in test

Signed-off-by: Anthony J Mirabella <a9@aneurysm9.com>

* receiver/prometheus: fix metric label tests

Signed-off-by: Anthony J Mirabella <a9@aneurysm9.com>

* receiver/prometheus: remove unused parameter from test helper

Signed-off-by: Anthony J Mirabella <a9@aneurysm9.com>

* make porto

Signed-off-by: Anthony J Mirabella <a9@aneurysm9.com>

bogdandrutu committed:
Remove legacy, unnecessary, .nocover files, used for coverage (#6462)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

dehaansa committed:
Add postgresql receiver Part 2 (#6154)

* Add postgresql receiver config and factory

* Address linter issues

* Fix issues with sample config

* Add makefile and go mods to postgresql

* Add postgresql receiver scraper

* Fix linting issues

* Add licenses

* Add license to receiver

* Update receiver/postgresqlreceiver/README.md

Co-authored-by: Daniel Jaglowski <dan.jaglowski@bluemedora.com>

* Refactor postgres config

* Update codeowners

* Update documentation

* Update receiver/postgresqlreceiver/testdata/config.yaml

Co-authored-by: Daniel Jaglowski <dan.jaglowski@bluemedora.com>

* Default databases config to empty slice rather than nil

* Refactor config tests to utilize default config

* initialize metadata to allow lint task to pass

* Update labels to attributes in metadata

* regenerate generated files

* Resolve missing factory implementation in postgres receiver

* Clean up postgres metadata generation

* Fix multimod verify

* Nest TLS settings under 'tls'

* refactor to use new config format

* Update expected metrics for test to be indented, remove anon function in client

* Resolve PR feedback by changing mocks to use mock.Mock

* Update lib pq dependency

* respond to PR feedback

* PR feedback to support partial errors and a few other points

* Respond to PR feedback

* Fix linter

* Update receiver/postgresqlreceiver/scraper.go

Co-authored-by: Daniel Jaglowski <dan.jaglowski@bluemedora.com>

* More PR feedback

* Simplify use of errors.AddPartial

* Fix linter issue

* Fix fix to linter issues

Co-authored-by: Daniel Jaglowski <dan.jaglowski@bluemedora.com>

dmitryax committed:
Change the way how metric type is specified in metadata.yaml (#6463)

To simplify the logic and make it simpler for users to read what properties are accepted for specific types.

tbarker25 committed:
`hostmetrics`: Upgrade to use gopsutil/v3 (#6468)

This removes the dependency on the older gopsutil library.

mx-psi committed:
[exporter/datadogexporter] Rely on http.Client's timeout instead of in exporterhelper's (#6414)

keep94 committed:
[tanzuobservability exporter] Add consumer for sum metrics. (#6385)

* [tanzuobservability exporter] Add consumer for sum metrics.

* Function headers to be more readable.

pushGaugeSingleNumberDataPoint->pushGaugeNumberDataPoint
pushSingleNumberDataPoint->pushNumberDataPoint

* Break verifySumConsumer into four separate tests.

* Minor changes to tests.

dmitryax committed:
[mdatagen] Extract test data yaml to files (#6470)

dmitryax committed:
[receiver/postgresql] Fix unit field in metadata.yaml (#6469)

dmitryax committed:
[mdatagen] Update metadata.yaml schema description (#6471)

Fix minor inconsistencies and update the schema according to the latest changes

Commit:
Google Cloud Spanner Receiver: added request tag label to metadata config for top query stats. (#6475)

pmalek committed:
filestorageextension: fix panic when configured directory cannot be accessed (#6103)

* filestorageextension: fix panic when configured directory cannot be accessed

* filestorageextension: move validation to config.Validate()

Commit:
Google Cloud Spanner Receiver: Added sample lock requests label to the top lock stats metrics. (#6466)

dmitryax committed:
[mdatagen] Fix metadata yaml in metricdata_test.go (#6473)

anneelisabethlelievre committed:
fix exemplars support (#6140)

* fix exemplars support

Signed-off-by: Anne-Elisabeth Lelièvre <anne-elisabeth.lelievre@ubisoft.com>

* switch test logger from zap.Nop to zaptest.

Signed-off-by: Anne-Elisabeth Lelièvre <anne-elisabeth.lelievre@ubisoft.com>

* fix lint error

Signed-off-by: Anne-Elisabeth Lelièvre <anne-elisabeth.lelievre@ubisoft.com>

dmitryax committed:
[mdatagen] Use path.Join instead of hardcoding testdata paths (#6472)

mx-psi committed:
[exporter/datadogexporter] Do not pick up localhost-like names from attributes (#6477)

We need to do this validation to exclude localhost-like names which are not really useful.

reese-lee committed:
Update README (#6481)

Reflects the change that our OTLP endpoint is GA

Commit:
`mysqlreceiver`: Remove requirement for password on MySQL (#6479)

* Remove requirement for password

* Remove invalid config tests

* Lint

* Update Readme

tigrannajaryan committed:
`hostmetricsreceiver`: Ensure SchemaURL is set (#6482)

This changes ensures that all metrics produced by hostmetricsreceiver
set the SchemaURL.

If no specific SchemaURL is set we assume all collected host metrics
confirm to our default SchemaURL. The assumption is that the code that
produces these metrics uses semantic conventions defined in
package "conventions".

mx-psi committed:
`datadogexporter`: Retry per network call (#6412)

* [exporter/datadogexporter] Refactor metadata retrier to take RetrySettings into account

* [exporter/datadogexporter] Do retries per network call on the metrics side

* Address review comment

* Fix lint

Commit:
`build`: produce darwin arm64 binaries (#6496)

This will produce binaries for darwin arm64.

djaglowski committed:
Reorganize code in scrapertest package (#6497)

The ability to compare metric related structs is very useful
in scraper testing, but it does not need to be the exclusive
purpose of this package. Anticipating the addition of other
capabilities, the files have been renamed to reflect their
specific purposes.

No functional changes are made here.

dmitryax committed:
[receiver/kubeletstats] Add support for read-only kubelet endpoint (#6488)

Some of the kubernetes distributions (e.g. GKE Autopilot) block access to secure kubelet endpoint which is exposed at 10250 port by default. Most of the metrics collected by the kubeletstats receiver can also be fetched from read-only kubelet endpoint which is exposed at 10255 port by default. But kubeletstats receiver's configuration interface expects only the secure endpoint, it adds https:// to the provided endpoint value and sets auth headers.

This adds ability to configure kubeletstats receiver to use read-only kubelet endpoint.

jmacd committed:
Statsdreceiver: remove several unnecessary fields (#6487)

Remove several fields from structs inside the statsdreceiver.

Simplifies existing statsdreceiver code, where there was redundancy. ​No functional changes.

The redundant labelKeys and labelValues are replaced with references to the attribute.Set, which contains the same data. The summary struct also had an extra copy of the metric name.

dmitryax committed:
Apply new experimental metrics builder for hostmetrics/cpu scraper (#6418)

Introduce new experimental metrics builder interface for metrics scrapers and apply it to hostmetrics/cpu scraper.

pmalek committed:
docs(filelogreceiver): add converter config doc (#6500)

BinaryFissionGames committed:
[internal/stanza] Batch in emitter instead of converter (#6378)

* Batch stanza entries up front

* Use same as previous default

* fix filelogreceiver tests

* refactor emitter.go to always defer mux unlocks

This makes it more clear/explicit what the critical sections are

* batchCount -> batchSize

* flatten resource aggregation code

* fix out of date comment for maxBatchSize

* Move checks for nil batch to avoid function calls

* Move emitter Stop directly after Start

* Use aggregation over batch for resource aggregation step

Also cleaned up comments and removed some unused bits here.

* Move converter resource aggregation map to reduced scope

Reducing the scope to just the aggregation loop prevents concurrent access.

* Remove reliance on internal/stanza defaults for filelogreceiver

* Remove paranoid nil check for cancel func

* default cancel func is noop for emitter

* Remove unnecessary option interface for emitter.

* Make sure to shutdown emitter in correct order.

* Combine some statements with following if in emitter

* Clarify batching logic in converter tests

* Fold some functions in emitter

Folded out makeNewBatch, makeNewBatchNoLock,
and appendEntry.

* add back in appendEntry, makeNewBatch

* Fix regressed if statement

mx-psi committed:
[exporter/datadogexporter] Take hostname into account for cache (#6223)

* [exporter/datadogexporter] Take hostname into account for cache

* Defensively copy tags on AddTags

* [exporter/datadogexporter] Fix m.host and add test for this

dmitryax committed:
[exporter/datadog] Add vanity import to fix build failures (#6502)

djaglowski committed:
Add djaglowski as code owner to several metric and log receivers (#6511)

* Add djaglowski as code owner to several metric and log receivers

* Update .github/CODEOWNERS

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

dmitryax committed:
[mdatagen] Fix experimental generator for metrics with no attributes (#6507)

Do not create/call <metric>DataPointsCapacity function for metrics with not attributes

djaglowski committed:
Separate scrapertest capabilities (#6499)

The scrapertest package currently has the ability to ignore
metric values while comparing metrics. This is useful for
integration testing, where exact values cannot necessarily be
anticipated.

This establishes an options pattern which allows use cases like
"ignore values" to be composed in a generic way. The ignore values
use case is now an option, and the implementation of the option
is based on masking the values by setting them to zero.

The masking capability here is specific to masking of values on
data points, but other similar capabilities can be added later.
In the long term, a composable set of such capabilities would
be extremely useful for testing purposes, and perhaps elsewhere.

bogdandrutu committed:
Update core to latest, fix breaking changes in config.Map constructors (#6515)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Use errors.Is to determine if an error is a specific type, for http/grpc stop errors (#6516)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Use errors.Is to determine if an error is a specific type, for http/grpc stop errors (#6516) (#6517)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

dmitryax committed:
Add @dmitryax as code owner to several components (#6524)

Splunk's exporters, couple orphaned receivers and internal/kubelet module

dmitryax committed:
[mdatagen] Add missed NumberDataPoints field to gauge metric type (#6521)

jamesmoessis committed:
Update extension/asapauth/README.md

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

jamesmoessis committed:
change package name asapauth -> asapauthextension

jamesmoessis committed:
change package name in CODEOWNERS

jamesmoessis committed:
add comments to fields in the example configuration for documentation

jamesmoessis committed:
upgrade to go-asap v2

jamesmoessis committed:
add type check closer to type definition

jamesmoessis committed:
move auth header provisioning and caching logic for RPC to PerRPCAuth()

jamesmoessis committed:
empty to reset build

jamesmoessis commented:
Not sure why the linter is failing: 

```
-package asapauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauthextension"
+package asapauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth"
```

Wants me to change the comment back to the old package name

mx-psi commented:
> Not sure why the linter is failing

You need to install porto and run `make goporto` (see #6498)

jpkrohling commented on .github/CODEOWNERS:
```diff
@@ -53,6 +53,7 @@ extension/httpforwarder/                             @open-telemetry/collector-c
 extension/observer/                                  @open-telemetry/collector-contrib-approvers @asuresh4 @jrcamp
 extension/bearertokenauthextension/                  @open-telemetry/collector-contrib-approvers @jpkrohling @pavankrish123
 extension/oauth2clientauthextension/                 @open-telemetry/collector-contrib-approvers @pavankrish123
+extension/asapauthextension                          @open-telemetry/collector-contrib-approvers @jamesmoessis @MovieStoreGuy
```
Could you please move it to its correct place when sorted alphabetically? I recently reordered the components on this file, you might need to rebase this PR.

jpkrohling commented on extension/asapauthextension/README.md:
```diff
@@ -0,0 +1,36 @@
+# ASAP Client Authentication Extension
+
+This extension provides [Atlassian Service Authentication Protocol](https://s2sauth.bitbucket.io/) (ASAP) client 
```
For my own education, do you have a resource with the differences between it and OAuth 2?

jpkrohling commented on extension/asapauthextension/config_test.go:
```diff
@@ -0,0 +1,93 @@
+
+// Test keys. Not for use anywhere but these tests.
+const (
+	PrivateKey = "data:application/pkcs8;kid=test;base64,MIIBUwIBADANBgkqhkiG9w0BAQEFAASCAT0wggE5AgEAAkEA0ZPr5JeyVDoB8RyZqQsx6qUD+9gMFg1/0hgdAvmytWBMXQJYdwkK2dFJwwZcWJVhJGcOJBDfB/8tcbdJd34KZQIDAQABAkBZD20tJTHJDSWKGsdJyNIbjqhUu4jXTkFFPK4Hd6jz3gV3fFvGnaolsD5Bt50dTXAiSCpFNSb9M9GY6XUAAdlBAiEA6MccfdZRfVapxKtAZbjXuAgMvnPtTvkVmwvhWLT5Wy0CIQDmfE8Et/pou0Jl6eM0eniT8/8oRzBWgy9ejDGfj86PGQIgWePqIL4OofRBgu0O5TlIN ...
```
Does this need to be exported?

jpkrohling commented on extension/asapauthextension/extension_test.go:
```diff
@@ -0,0 +1,114 @@
+	authHeaderValue := resp.Header.Get("Authorization")
+	tokenString := authHeaderValue[7:] // Remove prefix "Bearer "
+
+	validateAsapJwt(t, cfg, tokenString)
```
```suggestion
	validateAsapJWT(t, cfg, tokenString)
```

No need to fix this if this is the only thing pending for this PR, although it would be nice :-)

jpkrohling commented on extension/asapauthextension/go.mod:
```diff
@@ -0,0 +1,46 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth
```
This probably needs to have the `extension` suffix, which is likely the reason the lint check is failing.

```suggestion
module github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauthextension
```

jamesmoessis committed:
correct module name

jamesmoessis commented:
Something has gone... very wrong with this branch, I apoplogise for that. Currently fixing it up.

jamesmoessis commented:
I've messed up this branch, will raise a new PR. Sorry for the inconvenience
